src/pip/_internal/commands/cache.py

```diff
@@ -0,0 +1,152 @@
+        if not args or args[0] not in handlers:
+            logger.error("Need an action ({}) to perform.".format(
+                ", ".join(sorted(handlers)))
+            )
+            return ERROR
+
+        action = args[0]
+
+        # Error handling happens here, not in the action-handlers.
+        try:
+            handlers[action](options, args[1:])
+        except PipError as e:
+            logger.error(e.args[0])
+            return ERROR
+
+        return SUCCESS
+
+    def get_cache_info(self, options, args):
+        # type: (Values, List[Any]) -> None
+        num_packages = len(self._find_wheels(options, '*'))
+
+        cache_location = self._wheels_cache_dir(options)
+        cache_size = filesystem.friendly_directory_size(cache_location)
+
+        message = textwrap.dedent("""
+            Cache info:
+              Location: {location}
+              Packages: {package_count}
+              Size:     {size}
+        """).format(
+            location=cache_location,
+            package_count=num_packages,
+            size=cache_size,
+        ).strip()
+
+        logger.info(message)
+
+    def list_cache_items(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if len(args) > 1:
+            raise CommandError('Too many arguments')
+
+        if args:
+            pattern = args[0]
+        else:
+            pattern = '*'
+
+        files = self._find_wheels(options, pattern)
+
+        if not files:
+            logger.info('Nothing cached.')
+            return
+
+        results = []
+        for filename in files:
+            wheel = os.path.basename(filename)
+            size = filesystem.friendly_file_size(filename)
+            results.append(' - {} ({})'.format(wheel, size))
+        logger.info('Cache contents:\n')
+        logger.info('\n'.join(sorted(results)))
+
+    def remove_cache_items(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if len(args) > 1:
+            raise CommandError('Too many arguments')
+
+        if not args:
+            raise CommandError('Please provide a pattern')
+
+        files = self._find_wheels(options, args[0])
+        if not files:
+            raise CommandError('No matching packages')
+
+        for filename in files:
+            os.unlink(filename)
+            logger.debug('Removed %s', filename)
+        logger.info('Files removed: %s', len(files))
+
+    def purge_cache(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if args:
+            raise CommandError('Too many arguments')
+
+        return self.remove_cache_items(options, ['*'])
+
+    def _wheels_cache_dir(self, options):
+        # type: (Values) -> str
+        return os.path.join(options.cache_dir, 'wheels')
+
+    def _find_wheels(self, options, pattern):
+        # type: (Values, str) -> List[str]
+        wheel_dir = self._wheels_cache_dir(options)
+        return filesystem.find_files(wheel_dir, pattern + '*.whl')
```

Comment:
Maybe, we should allow the users to skip the "*" here -- which might be relevant if they want an exact match?

Reply:
For `list`, this is a relatively minor point, and I wouldn't be inclined to worry much. But in the case of `remove`, is prefix search what we want at all? Consider:

```
>.\.venv\Scripts\pip cache list
Cache contents:

 - MarkupSafe-1.1.1-cp38-cp38-win_amd64.whl (19 kB)
 - coverage-5.0.3-cp38-cp38-win_amd64.whl (205 kB)
 - distlib-0.3.0-cp38-none-any.whl (340 kB)
 - env_var_test-0.1-cp38-none-any.whl (1.3 kB)
 - env_var_test-0.1-cp38-none-any.whl (1.3 kB)
 - flit-2.2.0-cp38-none-any.whl (49 kB)
 - flit_core-2.2.0-cp38-none-any.whl (40 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - importlib_metadata-1.4.0-cp38-none-any.whl (29 kB)
 - msgpack-0.6.2-cp38-cp38-win_amd64.whl (14 kB)
 - pip-20.0.dev0-cp38-none-any.whl (1.4 MB)
 - pip-20.1.dev0-py2.py3-none-any.whl (1.4 MB)
 - pip-20.1.dev0-py2.py3-none-any.whl (1.5 MB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg512-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - progress-1.5-cp38-none-any.whl (8.1 kB)
 - proj-0.1-cp38-none-any.whl (1.2 kB)
 - pycparser-2.19-py2.py3-none-any.whl (111 kB)
 - pyrsistent-0.15.7-cp38-cp38-win_amd64.whl (71 kB)
 - retrying-1.3.3-cp38-none-any.whl (11 kB)
 - scripttest-1.3-cp38-none-any.whl (7.8 kB)
 - setuptools_scm-3.4.3-cp38-none-any.whl (26 kB)
 - tox-3.14.4.dev4+g581cc25.d20200116-cp38-none-any.whl (82 kB)
 - zipp-2.0.1-cp38-none-any.whl (4.2 kB)
PS 15:02 [Git: pr_6391] {00:00.289} C:\Work\Projects\pip
>.\.venv\Scripts\pip cache remove pkg
Files removed: 9
```

I'm pretty sure my instinct would have been that I meant to remove wheels for project `pkg`, not for `pkg123`, `pkg512`, or `pkgX`.

My inclination would be to make the argument a wildcard match on *just* the project name part of the wheel - specifically because of `remove`.

Reply:
I agree that for `pip cache remove`, this is definitely a problem. I'm not sure how to handle it, though &mdash; the main point of the `*` is to allow you to do (using one from your results as an example) e.g. `pip cache remove tox` or `pip cache remove tox-3.14.4` instead of `pip cache remove tox-3.14.4.dev4+g581cc25.d20200116-cp38-none-any.whl`.

I guess `-*.whl` would be an improvement, at least. But can the name include a hyphen in `.whl` files? E.g., if a package is called `foo-bar`, is the wheel `foo-bar-<version>-<...>.whl`? If so, this still leaves the possibility of incorrect matches (like `pip remove foo` removing both `foo` and `foo-bar`).

Reply:
Looking into this more, according to [PEP 427 -- The Wheel Binary Package Format 1.0](https://www.python.org/dev/peps/pep-0427/#file-name-convention):

> The wheel filename is `{distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-{platform tag}.whl`.

This means, after `{distribution}`, there's either 4 or 5 hyphens depending on if there's a build tag. I think I may need to use two regular expressions (one with the build tag, one without) and just check if each filename matches _either_ of them, and include the ones that do?

Reply:
The details of wheel names are [in the PEP](https://www.python.org/dev/peps/pep-0427/#file-name-convention), but no, the project name can never contain a `-` (non-alphanumerics are [canonicalised](https://www.python.org/dev/peps/pep-0427/#escaping-and-unicode) to `_`) so the only place you have `-` is separating the parts.

We could do something like `pattern + ("*.whl" if "-" in pattern else "-*.whl")` - this would mean that a pure project name is matched exactly, but if the user specifies `name-version`, the version is treated as a prefix.

This seems like it's sufficient - I don't think we need to make this over-complicated at this stage. If we hit issues, they can easily enough be fixed in a follow-up PR that tightens things up based on real use cases. That's better than trying to either guess what could go wrong, or over-engineer things from the start.

(Even the `pip cache remove` case isn't *that* serious, if we delete more than the user intended, it's only a cache for performance, and the wheels will get rebuilt anyway, so there's no major damage caused).

Reply:
@pfmoore I went ahead with the approach you suggested, and added a comment explaining the reasoning. And I agree, it's not a serious problem, but it's always nice for it to behave intuitively. Especially if the fix is this simple. :slightly_smiling_face: